### PR TITLE
feature/S4-37 - Select Signatories

### PR DESCRIPTION
**Pull request by peadarkelly**
Screenshots:

**Not a director selected:**
Screenshot 2020-07-06 at 15 45 57

**Both directors displayed:**
Screenshot 2020-07-06 at 15 46 05

**Error when no signatory is selected:**
Screenshot 2020-07-06 at 15 46 16

**Error when majority is not selected:**
Screenshot 2020-07-06 at 15 46 24

**Director is selected:**
Screenshot 2020-07-06 at 15 46 30

**Only one director is displayed:**
Screenshot 2020-07-06 at 15 46 37

### CODE REVIEW

File: src/controllers/selectDirector.controller.ts [hunk 1]
```diff
@@ -91,19 +91,30 @@ export class SelectDirectorController extends BaseController {
   }
 
   private updateSession(body: SelectDirectorFormModel, selectedDirector?: Optional<DirectorDetails>): void {
+    const session: DissolutionSession = this.session.getDissolutionSession(this.httpContext.request)!
+
+    if (!this.hasFormChanged(body, session)) {
```
Comment (peadarkelly) [inline src/controllers/selectDirector.controller.ts:96]: This has been agreed with Jayti. If the user has not changed anything on the page, then do not rewrite to session.

File: src/models/session/directorToSign.model.ts [hunk 2]
```diff
@@ -1,6 +1,7 @@
 export default interface DirectorToSign {
   id: string
   name: string
+  isApplicant: boolean
```
Comment (peadarkelly) [inline src/models/session/directorToSign.model.ts:4]: this was introduced to differentiate between the director selected on the selectDirector page vs the signatories selected on the selectSignatories page. Only the directors selected on the selectSignatories page will be shown on the define signatory info page.

File: src/views/select-director.njk [hunk 3]
```diff
@@ -6,7 +6,7 @@
 {% from 'govuk/components/button/macro.njk'        import govukButton %}
 
 {% block pageTitle %}
-  Tell us some details
+  Which director are you?
```
Comment (peadarkelly) [inline src/views/select-director.njk:9]: This was missed in a previous story

File: src/controllers/selectSignatories.controller.ts [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+    const signatories: DirectorDetails[] = await this.getSignatories(session.selectDirectorForm!.director!)
+
+    return this.renderView(signatories, session.selectSignatoriesForm)
+  }
+
+  @httpPost('')
+  public async post(@requestBody() body: SelectSignatoriesFormModel): Promise<string | RedirectResult> {
+    if (typeof body.signatories === 'string') {
+      body.signatories = [body.signatories]
+    }
```
Comment (peadarkelly) [inline src/controllers/selectSignatories.controller.ts:51]: this is a small niche of checkboxes. When you submit a single checkbox, it comes through as a string. When you submit multiple checkboxes, it comes through as a string[]. This is a small bit of code to ensure that we only need to treat them as string[] to reduce complexity.

File: src/services/company-officers/companyOfficers.service.ts [hunk 5]
```diff
@@ -16,16 +16,31 @@ export default class CompanyOfficersService {
...
-    return response.resource!.items
+    const activeDirectors: DirectorDetails[] = response.resource!.items
       .filter((officer: CompanyOfficer) => officer.officerRole === 'director')
       .filter((director: CompanyOfficer) => !director.resignedOn)
       .map((activeDirector: CompanyOfficer) => this.directorMapper.mapToDirectorDetails(activeDirector))
+
+    return directorToExclude ? this.exclueDirector(activeDirectors, directorToExclude) : activeDirectors
+  }
+
+  private exclueDirector(activeDirectors: DirectorDetails[], directorToExclude: string): DirectorDetails[] {
```
Comment (mateuszc-kainos) [inline src/services/company-officers/companyOfficers.service.ts:34]: exclude

File: src/services/company-officers/companyOfficers.service.ts [hunk 6]
```diff
@@ -16,16 +16,31 @@ export default class CompanyOfficersService {
...
+    return activeDirectors.filter(activeDirector => activeDirector.id !== directorToExclude)
+  }
+
+  public getMinimumNumberOfSignatores(totalSignatories: number, selectedDirector: string): number {
```
Comment (mateuszc-kainos) [inline src/services/company-officers/companyOfficers.service.ts:38]: signatories

File: src/services/company-officers/companyOfficers.service.ts [hunk 7]
```diff
@@ -16,16 +16,31 @@ export default class CompanyOfficersService {
...
+    const isApplicantADirector: boolean = selectedDirector !== 'other'
+    const totalActiveDirectors: number = isApplicantADirector ? totalSignatories + 1 : totalSignatories
+
+    const majority: number = Math.floor(((totalActiveDirectors / 2) + 1))
```
Comment (mateuszc-kainos) [inline src/services/company-officers/companyOfficers.service.ts:42]: do we need Math.floor here? I thought we were operating on integers here

Comment (peadarkelly) [reply src/services/company-officers/companyOfficers.service.ts:42]: Yup - JavaScript does not differentiate between ints, doubles and floats. They all come under the `number` type. Try it out in your chrome console to test the different values.

**Pull request merged**